vm/systemSmartContracts/freezeAccount.go

```diff
@@ -0,0 +1,284 @@
+
+// TODO:
+// 1. Add builtin function
+// 2. Move Guardian structs to elrond-go-core
+
+// Key prefixes
+const (
+	GuardiansKey = "guardians"
+)
+
+// Functions
+const (
+	setGuardian = "setGuardian"
+)
+
+type Guardian struct {
+	Address         []byte
+	ActivationEpoch uint32
+}
+
+type Guardians struct {
+	Data []*Guardian
+}
+
+type ArgsFreezeAccountSC struct {
+	GasCost              vm.GasCost
+	Marshaller           marshal.Marshalizer
+	SystemEI             vm.SystemEI
+	PubKeyConverter      core.PubkeyConverter
+	AccountFreezerConfig config.AccountFreezerSCConfig
+	EpochConfig          config.EpochConfig
+	EpochNotifier        vm.EpochNotifier
+}
+
+type freezeAccount struct {
+	gasCost              vm.GasCost
+	marshaller           marshal.Marshalizer
+	systemEI             vm.SystemEI
+	pubKeyConverter      core.PubkeyConverter
+	guardianEnableEpochs uint32
+	mutExecution         sync.RWMutex
+
+	enableEpoch uint32
+	flagEnabled atomic.Flag
+}
+
+func NewFreezeAccountSmartContract(args ArgsFreezeAccountSC) (*freezeAccount, error) {
+	if check.IfNil(args.Marshaller) {
+		return nil, core.ErrNilMarshalizer
+	}
+	if check.IfNil(args.SystemEI) {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+	if check.IfNil(args.PubKeyConverter) {
+		return nil, vm.ErrNilAddressPubKeyConverter
+	}
+
+	accountFreezer := &freezeAccount{
+		gasCost:              args.GasCost,
+		marshaller:           args.Marshaller,
+		systemEI:             args.SystemEI,
+		pubKeyConverter:      args.PubKeyConverter,
+		guardianEnableEpochs: args.AccountFreezerConfig.GuardianEnableEpochs,
+		enableEpoch:          args.EpochConfig.EnableEpochs.AccountFreezerEnableEpoch,
+		mutExecution:         sync.RWMutex{},
+	}
+	logAccountFreezer.Debug("account freezer enable epoch", accountFreezer.enableEpoch)
+	args.EpochNotifier.RegisterNotifyHandler(accountFreezer)
+
+	return accountFreezer, nil
+}
+
+func (fa *freezeAccount) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	fa.mutExecution.RLock()
+	defer fa.mutExecution.RUnlock()
+
+	if CheckIfNil(args) != nil {
+		return vmcommon.UserError
+	}
+	if !fa.flagEnabled.IsSet() {
+		fa.systemEI.AddReturnMessage(fmt.Sprintf("account freezer not enabled yet, enable epoch: %d", fa.enableEpoch))
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case setGuardian:
+		return fa.setGuardian(args)
+	default:
+		return vmcommon.UserError
+	}
+}
+
+// todo; check if guardian is already stored?
+
+// 1. User does NOT have any guardian => set guardian
+// 2. User has ONE guardian pending => does not set, wait until first one is set
+// 3. User has ONE guardian enabled => add it
+// 4. User has TWO guardians. FIRST is enabled, SECOND is pending => change pending with current one / does nothing until it is set
+// 5. User has TWO guardians. FIRST is enabled, SECOND is enabled => replace oldest one
+func (fa *freezeAccount) setGuardian(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```

Comment:
This is a built in function - not a system smart contract. System SCs are only on metachain.

Reply:
Thanks for pointing out. Will move this code to `elrond-vm-common`.